bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+      bindBindCommon(bind, sym, right)
+      
+    case CallStatement(callable, args) =>
+      walkExpr(callable)
+      args map walkExpr
+      for ((sym, decl) <- declVars if decl.state == Captured) {
+        tagVariable(decl.v, ToInstantiate)
+      }
+      
+    case IfStatement(cond, trueStat, elseStat) =>
+      walkExpr(cond)
+      conjugate(declVars, _.walkStat(trueStat), _.walkStat(elseStat))
+      
+    case MatchStatement(value, clauses, elseStatement) =>
+      walkExpr(value)
+      conjugate(declVars, (clauses.map {
+        clause =>
+          m: OnStackMarking =>
+           clause.guard foreach {g => m.walkExpr(g)}
+           m.walkStat(clause.body)
+      } :+ {m: OnStackMarking => m.walkStat(elseStatement)}) :_*)
+      
+    case TryStatement(body, exceptionVar, catchBody) =>
+      affectedVariables(null, { m =>
+        m.walkStat(body)
+        m.walkStat(catchBody)
+      }) foreach (tagVariable(_, ToInstantiate))
+      
+    case _ => super.walkStat(statement)
+  }
+  
+  override def walkExpr(expression: Expression): Unit = expression match {
+    case bind @ BindExpression(left @ Variable(sym), right) =>
+      bindBindCommon(bind, sym, right)
+      
+    case Variable(sym) =>
+      for (decl <- declVars.get(sym) if !decl.state.locked) {
+        tagVariable(decl.v, ToInstantiate)
+      }
+      
+    case ProcExpression(name, args, body, flags) =>
+      affectedVariables(null, { m =>
+        m.walkStat(body)
+      }) foreach (tagVariable(_, Captured))
+      
+    case IfExpression(cond, trueExpr, elseExpr) =>
+      walkExpr(cond)
+      conjugate(declVars, _.walkExpr(trueExpr), _.walkExpr(elseExpr))
+      
+    case MatchExpression(value, clauses, elseExpression) =>
+      walkExpr(value)
+      conjugate(declVars, (clauses.map { clause =>
+        m: OnStackMarking =>
+         clause.guard foreach {g => m.walkExpr(g)}
+         m.walkExpr(clause.body)
+      } :+ {m: OnStackMarking => m.walkExpr(elseExpression)}):_*)
+      
+    case TryExpression(body, exceptionVar, catchBody) =>
+      affectedVariables(null, { m =>
+        m.walkExpr(body)
+        m.walkExpr(catchBody)
+      }) foreach (tagVariable(_, ToInstantiate))
+      
+    case ShortCircuitBinaryOp(left, op, right) =>
+      affectedVariables(null, { m =>
+        m.walkExpr(left)
+        m.walkExpr(right)
+      }) foreach (tagVariable(_, ToInstantiate))
+      
+    case _ => super.walkExpr(expression)
+  }
+  
+  def inspectLocalCommon(decls: Seq[Variable], statOrExpr: StatOrExpr) = {
+      for (decl <- decls if !decl.symbol.isCapture) {
+        declVars.put(decl.symbol, VarAndState(decl, Declared))
+      }
+      statOrExpr match {
+        case stat: Statement  => walkStat(stat)
+        case expr: Expression => walkExpr(expr)
+      }
+      for (decl <- decls if declVars.contains(decl.symbol)) {
+        if (declVars.get(decl.symbol).get.state.onStack) {
+        	decl.onStack = true
```

Comment:
So Variable.onStack is only used to store the final state so the `local` knows which vars it should instantiate, right?